src/handler/mod.rs

```diff
@@ -480,7 +467,26 @@ impl Handler {
 
         let (packet, initiating_session) = {
             if let Some(session) = self.sessions.get_mut(&node_address) {
-                // Encrypt the message and send
+                // Encrypt the message and send unless it is a hole-punch PING that already has a
+                // session. This would be the case if this node is the initiator and it is not
+                // behind a NAT or is port-forwarded (only receiver is behind NAT), then the PING
+                // from the receiver which according to the NAT traversal protocol is sent before
+                // the PING from the initiator, has already lead to an established session between
+                // the initiator and the receiver.
+                if hole_punch {
+                    let request_id = request.id;
+                    if let Err(e) = self
+                        .service_send
+                        .send(HandlerOut::RequestFailed(
+                            request_id,
+                            RequestError::UnusedHolePunchPingToReceiver,
+                        ))
+                        .await
+                    {
+                        warn!("Failed to inform request failure {}", e)
+                    }
+                    return Ok(());
+                }
```

Comment:
I'm trying to figure out why is this done here. This function is `send_request`, which, when hole punching, completely ignores the request and instead informs an error?

Reply:
so if there is an ongoing challenge or request, this ping is queued as pending. It seems to be there is no need for that?